ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ObjectDiffusion/Smoke.hs

```diff
@@ -0,0 +1,306 @@
+          []
+          objects
+
+instance WithId SmokeObject SmokeObjectId where getId = getSmokeObjectId
+
+{-------------------------------------------------------------------------------
+  Mock objectPools
+-------------------------------------------------------------------------------}
+
+newtype SmokeObjectId = SmokeObjectId Int
+  deriving (Eq, Ord, Show, NoThunks, Arbitrary)
+
+newtype SmokeObject = SmokeObject {getSmokeObjectId :: SmokeObjectId}
+  deriving (Eq, Ord, Show, NoThunks, Arbitrary)
+
+newtype SmokeObjectPool m = SmokeObjectPool (StrictTVar m [SmokeObject])
+
+newObjectPool :: MonadSTM m => [SmokeObject] -> m (SmokeObjectPool m)
+newObjectPool initialPoolContent = SmokeObjectPool <$> uncheckedNewTVarM initialPoolContent
+
+makeObjectPoolReader ::
+  MonadSTM m => SmokeObjectPool m -> ObjectPoolReader SmokeObjectId SmokeObject Int m
+makeObjectPoolReader (SmokeObjectPool poolContentTvar) =
+  ObjectPoolReader
+    { oprObjectId = getSmokeObjectId
+    , oprObjectsAfter = \minTicketNo limit -> do
+        poolContent <- readTVar poolContentTvar
+        pure $
+          take (fromIntegral limit) $
+            drop (minTicketNo + 1) $
+              ( (\(ticketNo, smokeObject) -> (ticketNo, getSmokeObjectId smokeObject, pure smokeObject))
+                  <$> zip [(0 :: Int) ..] poolContent
+              )
+    , oprZeroTicketNo = -1 -- objectPoolObjectIdsAfter uses strict comparison, and first ticketNo is 0.
+    }
+
+makeObjectPoolWriter ::
+  MonadSTM m => SmokeObjectPool m -> ObjectPoolWriter SmokeObjectId SmokeObject m
+makeObjectPoolWriter (SmokeObjectPool poolContentTvar) =
+  ObjectPoolWriter
+    { opwObjectId = getSmokeObjectId
+    , opwAddObjects = \objects -> do
+        atomically $ modifyTVar poolContentTvar (++ objects)
+        return ()
+    , opwHasObject = do
+        poolContent <- atomically $ readTVar poolContentTvar
+        pure $ \objectId -> any (\obj -> getSmokeObjectId obj == objectId) poolContent
+    }
+
+mkMockPoolInterfaces ::
+  MonadSTM m =>
+  [SmokeObject] ->
+  m
+    ( ObjectPoolReader SmokeObjectId SmokeObject Int m
+    , ObjectPoolWriter SmokeObjectId SmokeObject m
+    , m [SmokeObject]
+    )
+mkMockPoolInterfaces objects = do
+  outboundPool <- newObjectPool objects
+  inboundPool@(SmokeObjectPool tvar) <- newObjectPool []
+
+  let outboundPoolReader = makeObjectPoolReader outboundPool
+      inboundPoolWriter = makeObjectPoolWriter inboundPool
+
+  return (outboundPoolReader, inboundPoolWriter, atomically $ readTVar tvar)
+
+{-------------------------------------------------------------------------------
+  Main properties
+-------------------------------------------------------------------------------}
+
+-- Protocol constants
+
+maxFifoSize :: NumObjectsOutstanding
+maxFifoSize = NumObjectsOutstanding 10
+
+maxIdsToReq :: NumObjectIdsReq
+maxIdsToReq = NumObjectIdsReq 3
+
+maxObjectsToReq :: NumObjectsReq
+maxObjectsToReq = NumObjectsReq 2
```

Comment:
What about randomizing on those too?

Reply:
Usually, unless there is a significant complication (which I don't expect), it is nice to randomly generate these to have slightly fewer magic hardcoded constants :smile:

Reply:
Yeah, it's very cheap and will cover more test cases. Just add simple arbitrary instances picking ints between two somewhat reasonable values, and add these as arguments of your property and you're done!

Reply:
Done in fbab9259b4f312125efc17624a104536173d7431, although my constraints are very arbitrary. Please review.

Reply:
This is fine, there isn't a principled approach here; essentially all property tests must choose more or less arbitrary constants here. The only thing one could do would be to scale them via eg [`sized`](https://hackage-content.haskell.org/package/QuickCheck-2.16.0.0/docs/Test-QuickCheck.html#v:sized), but that is not blocking this PR.